SalesforceCommerceCloud/pwa-kit
# [PWA-Kit Extensibility] Implement wild-card module resolution logic @W-15932154

# Description

The PWA-Kit platform uses `babel-node` to run our PWA-Kit applications (think, typescript-minimal and retail-react-app). Because of this, we have 2 distinct places in our code where it's required that we modify the module resolution algorithm to support extensibility overrides. The first place is via a **webpack** plugin and the other is the **babel-node** plugin.

The logic used by these plugins is for the most part the same, so it's important that we don't repeat ourselves and pull that logic out into a set of reusable utilities.  

So, in this PR we create those utilities to be used by a webpack and a babel plugin in order to implement wildcard import statement syntax. What this syntax allows us to do is, import a module from a known list of packages where the modules being imported is known to exist in at least one of the packages. 

This is the basis of "multi-extensibility" and those "known list" of packages are in fact the list of extensions. 

# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- Create utils for building resolves "packageIterator" array

# How to Test-Drive This PR

- Run tests, test should pass.

- Alternately I've deployed a version of the typescript-minimal application that implemented a webpack plugin that uses the core logic in this pr. You can see it [here](https://github.com/SalesforceCommerceCloud/pwa-kit/compare/bendvc/W-15932154-demo#diff-a5cdf6ba9e430099e2e2e85c652f1df6c4554f12207a5c3dee81936fa40cbfeb), you can view pages that have been added via extensions [here](https://scaffold-pwa-extensibility.mobify-storefront.com/test-a), [here](https://scaffold-pwa-extensibility.mobify-storefront.com/test-b) and [here](https://scaffold-pwa-extensibility.mobify-storefront.com/test-c).

- If you are interested in seeing how the core logic will be used you can take a look at the following branch comparison [here](https://github.com/SalesforceCommerceCloud/pwa-kit/compare/bendvc/W-15932154-demo...bendvc/W-15932154)

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
